jitbuilder/lljb/src/IRVisitor.cpp

```diff
@@ -0,0 +1,559 @@
+      {
+      llvm::StructType * llvmStruct = llvm::dyn_cast<llvm::StructType>(I.getAllocatedType());
+      ilValue = _builder->CreateLocalStruct(_methodBuilder->getIlType(_td,llvmStruct));
+      unsigned elIndex = 0;
+      for (auto elIter = llvmStruct->element_begin();
+            elIter != llvmStruct->element_end(); ++elIter)
+         {
+         llvm::Type * fieldType = llvmStruct->getElementType(elIndex);
+         if (fieldType->isArrayTy())
+            {
+            TR::IlValue * arrayField = _builder->CreateLocalArray(
+               fieldType->getArrayNumElements(),
+               _methodBuilder->getIlType(_td,fieldType));
+            _builder->StoreAt(
+               _builder->StructFieldInstanceAddress(
+                  llvmStruct->getStructName().data(),
+                  _methodBuilder->getMemberNameFromIndex(elIndex),
+                  ilValue),
+               arrayField);
+            }
+         elIndex++;
+         }
+      }
+   else if (I.getAllocatedType()->isArrayTy())
+      {
+      llvm::Type * type = I.getAllocatedType();
+      unsigned numElements = type->getArrayNumElements();
+      type = type->getArrayElementType();
+      while (1)
+         {
+         if (type->isArrayTy())
+            {
+            numElements *= type->getArrayNumElements();
+            type = type->getArrayElementType();
+            }
+         else
+            {
+            break;
+            }
+         }
+      ilValue = _builder->CreateLocalArray(
+                           numElements,
+                           _methodBuilder->getIlType(_td,type));
+      }
+   else
+      {
+      _methodBuilder->allocateLocal(
+      &I,
+      _methodBuilder->getIlType(_td, I.getAllocatedType()));
+      }
+   _methodBuilder->mapIRtoIlValue(&I, ilValue);
+   }
+
+void
+IRVisitor::visitLoadInst(llvm::LoadInst &I)
+   {
+   llvm::Value * source = I.getPointerOperand();
+   TR::IlValue * loadedVal = nullptr;
+   if (_methodBuilder->isIndirectLoadOrStore(source))
+      {
+      TR::IlValue * ilSrc = getIlValue(source);
+      loadedVal = _builder->LoadAt(
+                              _td->PointerTo(_methodBuilder->getIlType(_td,I.getType())),
+                              _builder->
+                              IndexAt(
+                                 _td->PointerTo(_methodBuilder->getIlType(_td,I.getType())),
+                                 ilSrc,
+                                 _builder->ConstInt32(0)));
+      }
+   else
+      {
+      loadedVal = _builder->Load(_methodBuilder->getLocalNameFromValue(source));
+      }
+   _methodBuilder->mapIRtoIlValue(&I, loadedVal);
+   }
+
+void
+IRVisitor::visitStoreInst(llvm::StoreInst &I)
+   {
+   llvm::Value * dest = I.getPointerOperand();
+   llvm::Value * value = I.getOperand(0);
+   TR::IlValue * ilValue = getIlValue(value);
+   if (_methodBuilder->isIndirectLoadOrStore(dest))
+      {
+      _builder->StoreAt(
+               _builder->IndexAt(
```

Comment:
Same "IndexAt" comment here...if the index is 0, I wouldn't have expected you need it.

Reply:
See reply to the other comment.